### PR TITLE
Allow data to be copied from Mathesar and pasted elsewhere

### LINKED ISSUE #1688: Allow content in selected cells to be copied and pasted elsewhere
We should allow selected cells from Tables or other tabular views to be copied and pasted elsewhere.

### PR DESCRIPTION
**Pull request by seancolsen**
Fixes #1688

## Notes

- This PR provides a partial implementation of the "copy" functionality that I've [proposed](https://github.com/centerofci/mathesar-wiki/pull/90) in the [Clipboard Interaction Specs](https://github.com/centerofci/mathesar-wiki/blob/clipboard-specs/design/specs/clipboard-interactions.md). **It would be good to read the spec before reviewing this PR.**

- Scheduling of review

    - If we had ample time, I'd be inclined to get the Clipboard Interaction specs merged before merging this partial implementation. Instead, I'm doing the two in parallel. Why? Because I actually started with the implementation in order to get a good sense of the available browser APIs and their limitations. Getting my hands dirty with this first part of the implementation made the spec process much clearer than it would have been if I had written the specs before writing any code. I'm hoping that the "copy" portion of the spec is sufficiently straightforward that the code in this PR will require minimal alteration after final approval of the spec. So I think it's best to go ahead and try getting this merged as-is even though we're still hammering out the spec.

- In this PR:

    - You can copy data from Mathesar as TSV plain text. This allows you to paste it into other applications like LibreOffice Calc, Google Sheets, and Airtable.

    - You can copy cells from a table sheet or an exploration sheet.

    - The cell values are copied as _formatted values_. This means that FK cells are copied as their record summaries, data cells are copied as their formatted dates and so on.

- What's missing:

    - This PR does not implement any paste functionality.

    - This PR does not write the Mathesar-specific representation to the clipboard when copying. That's okay because we haven't implemented paste yet.

    - This PR does not have adequate handling of newlines/tab characters in cell content. As noted in the spec, I'm interested in utilizing a 3rd party lib for that, so I'd like to do a bit more research before going into that implementation. I'll track that work in tickets after merging this PR and merging the design spec.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by seancolsen** [comment]
@rajatvijay I assigned this to you because you asked 1/1 if you could review this.

@pavish I assigned this to you because you're marked as a reviewer on the [usability improvements project](https://wiki.mathesar.org/en/projects/2023-04-usability-improvements.md). There is also some of your code that I've changed which you might like to look over (e.g. `arrayFactory.ts` and related). But if you want to defer review to Rajat, that's fine with me.

**Review by pavish — CHANGES_REQUESTED** [review]
@seancolsen 

I tried testing this and I noticed that the copy functionality isn't working in Chrome and Safari. It's working in Firefox.

In Safari, selecting cells seems to also select content. I assume `user-select:none` isn't working, I'm not sure, I haven't dug deeper. Either way, the user selected content is being copied due to browser default action, but the selected cells content isn't copied.

I haven't reviewed the PR yet.

**Comment by seancolsen** [comment]
@pavish thanks for catching that.

> In Safari, selecting cells seems to also select content.

I'm reproducing this in the latest `develop`, so I opened #2806 for it.

I'll work on the issue with copy not behaving correctly in Chrome/Safari.

File: mathesar_ui/src/components/sheet/SheetClipboardHandler.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  strategy: CopyingStrategy,
+  recordSummaries: RecordSummariesForSheet,
+): string {
+  const record = indexedRecords.get(rowId);
+  if (!record) {
+    return '';
+  }
+  const rawCellValue: unknown = record[String(columnId)];
+  if (rawCellValue === undefined || rawCellValue === null) {
+    return '';
```
Comment (pavish) [inline mathesar_ui/src/components/sheet/SheetClipboardHandler.ts:43]: We should probably document `null` values would be copied as `''` in the clipboard specs.

File: mathesar_ui/src/components/cell-fabric/data-types/arrayFactory.ts [hunk 2]
```diff
@@ -17,45 +17,60 @@ export interface ArrayLikeColumn extends CellColumnLike {
...
-
-export default arrayType;
+    }),
+    getInput: (): ComponentAndProps => ({ component: TextInput }),
+    getDisplayFormatter: (column: ArrayLikeColumn) => {
+      const formatOneValue = makeDisplayFormatter(componentFactoryMap, column);
+      return (cellValue: unknown) => {
+        if (Array.isArray(cellValue)) {
+          return cellValue.map(formatOneValue).join(', ');
+        }
```
Comment (pavish) [inline mathesar_ui/src/components/cell-fabric/data-types/arrayFactory.ts:71]: The values can contain commas. We need to quote values here and escape quotes within the strings.

File: mathesar_ui/src/components/cell-fabric/data-types/string.ts [hunk 3]
```diff
@@ -44,6 +44,7 @@ const stringType: CellComponentFactory = {
       },
     };
   },
+  getDisplayFormatter: () => String,
```
Comment (pavish) [inline mathesar_ui/src/components/cell-fabric/data-types/string.ts:47]: We need to quote strings which have `\t` characters. This applies for all types.

**Review by pavish — APPROVED** [review]
@seancolsen Looks good overall, nice work on this.

There are a couple concerns:
* This still doesn't work on Safari. While the text selection bug is still there, the columns which are selected aren't being copied. A single active cell isn't being copied as well. 
   - Please create an issue to track this.
* We need to quote the values wherever necessary, and when we do that, we have to make sure we escape the quotes within them. I saw the comment on the description, please create a separate issue to track this.

Please go ahead and merge once the issues are created.

**Comment by pavish** [comment]
Followups:
* https://github.com/centerofci/mathesar/issues/2811
* https://github.com/centerofci/mathesar/issues/2812
* https://github.com/centerofci/mathesar/issues/2806

From @seancolsen:

> The bug with the cell selection and text selection is making it harder to test the clipboard feature. The clipboard feature still works in Safari though. Click one cell to select a cell. Then press an arrow key to select an adjacent cell. Then use your keyboard shortcut to copy the cell value. You should see a toast message at that point.

Fixing https://github.com/centerofci/mathesar/issues/2806 should fix the issue with copying in Safari.

**Pull request merged**